token-syncer/src/token_syncer/commands/cleanup.clj

```diff
@@ -0,0 +1,63 @@
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns token-syncer.commands.cleanup
+  (:require [clj-time.core :as t]
+            [clojure.tools.logging :as log]
+            [token-syncer.utils :as utils]))
+
+(defn cleanup-tokens
+  "Reads the latest tokens from the specified cluster and hard-deletes all the soft-deleted tokens last
+   updated before the specified time."
+  [{:keys [hard-delete-token load-token-list]} cluster-url before-epoch-time]
+  (let [index-entries (load-token-list cluster-url)
+        soft-deleted-index-entries (filter (fn [{:strs [deleted last-update-time]}]
+                                             (and (true? deleted)
+                                                  (< (utils/iso8601->millis last-update-time) before-epoch-time)))
+                                           index-entries)]
+    (log/info "found" (count soft-deleted-index-entries) "qualifying soft-deleted token(s) out of"
+              (count index-entries) "total token(s) on" cluster-url)
+    (doseq [{:strs [etag last-update-time token]} soft-deleted-index-entries]
+      (log/info "hard-deleting token" token "with etag" etag "last updated on" last-update-time)
+      (hard-delete-token cluster-url token etag))
+    (log/info "hard-deleted" (count soft-deleted-index-entries) "token(s) on cluster" cluster-url)
+    (->> soft-deleted-index-entries
+      (map #(get % "token"))
+      (into #{}))))
+
+(def cleanup-tokens-config
+  {:execute-command (fn execute-cleanup-tokens-command
+                      [{:keys [waiter-api]} {:keys [options]} arguments]
+                      (let [{:keys [before]} options]
+                        (if-not (= (count arguments) 1)
+                          {:exit-code 1
+                           :message (str "expected one argument URL, provided " (count arguments) ": " (vec arguments))}
+                          (let [before-time (if before
+                                              (Long/parseLong before)
+                                              (- (System/currentTimeMillis) (-> 1 t/weeks t/in-millis)))
+                                cluster-url (first arguments)]
+                            (log/info "hard-deleting tokens from" cluster-url "soft-deleted before"
+                                      (utils/millis->iso8601 before-time))
+                            (cleanup-tokens waiter-api cluster-url before-time)
+                            {:exit-code 0
+                             :message "exiting with code 0"}))))
+   :option-specs [["-b" "--before"
+                   (str "Hard-delete tokens soft-deleted before the specified time. "
+                        "Must be specified in as an epoch time. "
```

Comment:
Should we use iso8601 as the input format?

Reply:
Changed to use iso8601 as input format.